CryptoPrices/Sources/Data/Sources/NetworkCore/CoinAPIProtocol.swift

```diff
@@ -5,11 +5,14 @@
 //  Created by Doan Thieu on 01/12/2022.
 //
 
+import Entities
+
 // sourcery: AutoMockable
 public protocol CoinAPIProtocol {
 
     // TODO: Update when implement
     func myCoins() async throws -> [APICoin]
     func trendingCoins(coinIDs: [String]) async throws -> [APICoin]
     func coinDetail() async throws -> APICoin
+    func getChartPrices(coinID: String, filter: TimeFilter) async throws -> APIPrices
```

Comment:
> I think we can still make the network layer work without depending on the Domain layer here.

-> To be honest with you, please correct me if I am wrong but I am quite confused with this claim because we are already making the `Data` layer depend on the `Domain` layer, no? Practically, it is impossible to completely separate them and I actually feel it is natural to let the `Data` layer depend on the `Domain` layer. 🙏
For example, we have `RepositoryProtocol`  inside the `Domain` layer, and Repositories in the `Data` layer need to conform to the protocols, right?
![Screenshot 2022-12-28 at 15 31 06](https://user-images.githubusercontent.com/70877098/209783438-d323d039-ecdf-4024-939e-db92aaa55c10.png)

> We just provide directly a `numberOfDays`, we don't have to use higher abstraction like a custom enum here. The parameter needs update too.

-> In the beginning, I had the same thought as you to just provide `numberOfDays` as `Int` too 👍 Then, when thinking in the long run, the main purpose of introducing `enum` here is for reusing it in the future in the logic layer - `ViewModel` or even the `View` layer (since we are using SwiftUI). Later on, when we need to switch between different filters to call the correct API and draw different charts, now we already have the enum representation to work with. And the day's calculation introduced in the enum is just a technique to group related code for a more descriptive way of usage. 🙏

> The function name, parameter types, and return types between different layers are not necessarily the same and actually have nothing to do with others.

-> It is my intention to use parameter types as `enum` for a more descriptive type of writing code instead of just providing plain numbers and we have to add extra comments/descriptions for it at a later time in the Network layer here for protocols. Please correct me but I still don't think there is a valid concern to stop using `enum` at this point.   🙇‍♂️ Why do we have to stick with the `low-level`, `raw data` type usage for protocols? 🤔

Reply:
@minhnimble Thank you for your detailed feedback 🙏  I'd like to continue sharing my concerns here:
- About using the `TimeFilter` enum: I think you missed my point. I didn't say to not use the enum, but rather the opposite:
  > So it's great that on the Domain layer we can provide the TimeFilter type that abstracts away the number of days with a set of predefined ones.

  You're encouraged to use it in the higher layers like `Repository`, `UseCase`, or even better in the `ViewModel` and `View` which are closer to the presentation layer. What I meant was you shouldn't use that enum in the network layer. Here're 2 obvious reasons:
  - The `CoinAPIProtocol` currently can do `func getChartPrices(coinID: String, filter: TimeFilter) async throws -> APIPrices`. Meaning that it _only can make the request with the days that are one of the values 1, 7, 30, 365, and 1285_. What if I want to get the history price for 3, 57, 2589 days, or any other given day? I can't, right? By forcing it to accept only a small set of values, you limit the network layer's functionality and capability.
  - Using the enum here not only limits the network's capability, but it also becomes too coupled with the app business. Since the `MyCoinView` display the chart with a predefined set of days, you modelize the network in a way to support just that.
 
    What if the `MyCoinView` can also display data for like ten years? You have to add a new case for `TimeFilter` and thus have to update the network 🤔?

    Another example is that suppose our app allows users to register an account, with 2 tiers: free and premium. A free account can only see the historical prices for what is defined in `TimeFilter`. A premium account has an additional option to choose whatever day ranges they want to see (for e.g with a Slider on the UI with a value between 1 and 10_000). In that case, updating `ViewModel`, and `UseCase` is logical and necessary. But modify the network layer to make it work? It's supposed to be updated only on some rare cases like CoinGecko API breaks something, or we update from `APIv1` to `APIv2`, etc,... But it _never updates when the app business changes, or when the UI/UX changes_. You have to agree with me it's strange if it would be the case right 🙏

   As a side note, we can use the enum in the network layer when the CoinGecko API accepts those, and a very good candidate is the `vs_currency` parameter. They accept values like `usd`, `eur`, `jpy`, etc, ... They must keep a valid set on the database schema so the caller can't just make a request with an arbitrary string like `lorem ipsum`.

- About the relationship between layers, the `Data` _does_ depend on the `Domain` in some senses, but it's more complicated than that. In SPM terms, the _package_ is just kind of a container (it may be oversimplified but I think you got the point). What is matter here is the modules that the package exposed. So the `Data` _package_ has 4 _modules_:
  - `NetworkCore`
  - `NetworkExtension`
  - `Repositories`
  - `DataTestHelpers`
  
  while the `Domain` _package_ has following _modules_:
  - `Entities`
  - `RepositoryProtocol`
  - `UseCaseProtocol`
  - `UseCases`
  - `DomainTestHelpers`

  As you indicated, the `Repositories` module (concrete type) of the `Data` depends on the `RepositoryProtocol` (interface) of the `Domain`. What I meant when I say `can still make the network layer work without depending on the Domain layer` is that you don't have to make the `NetworkExtension` depend on the `Entities` (to use the `TimeFilter`). Of course, you can but with the ☝️ above point, you shouldn't and don't have to.

Last but not least, I'd like to take this chance to present a clear _picture_ of how we structure and architect the app so far. Why? Because a picture is worth a thousand words 🙏 :

![CryptoPricesArchitecture](https://user-images.githubusercontent.com/17972674/209901107-aeb5d79e-a0d9-44b2-8ea9-78eb57289fdb.png)

(Don't judge my drawing skills. Don't even ask me about UML stuff either 😄 I'm very bad at those kinds of things)
We modulize the app by distributing the sources into different layers. Each layer contains one or more packages. Each package contains one or more modules that can be `import`ed by others.
Some worth-to-mention points here (please take your time to check this later, and feel free to raise you ideas):
- Layers are organized vertically. A module of a lower layer depends on one or more modules of the higher layer, but not vice versa.
- Modules from packages on the same layer don't know each other. For e.g. the `Home` vs the `MyCoin` in the __Features__. Or the `BuildTools` vs the `Styleguide` in the __Utilities__.
- A concrete type of module on a lower layer depends on the abstraction (interface) of a higher layer. For e.g. the `Repositories` (an implementation of the repository) depends on the `NetworkCore`. So when we change the implementation of the network (`NetworkExtension`),  say we switch from `Pilot` to `Moya` and `Alamofire`, that `Repositories` won't change and won't be re-built when the app builds. And you definitely know that this has the benefit for unit testing too 🙏 
- The app is the one that uses mostly concrete types. This is obvious because the implementation is what makes the app does what it advertises (business logic)

Reply:
@Thieurom with the example of different account tiers, I think it starts to make more sense to me now. 👍I think you can just start by explaining with visual examples next time for such architectural ideas like this 🙈

-> I will remove the enum usage for the network layer i.e. `CoinAPIProtocol` 🚀

Reply:
Thank you, @Thieurom, for the detailed explanation with a great visual example 🙏 and @minhnimble for raising this, which helps me clearer in this architecture 🙏.

Reply:
@minhnimble  @nkhanh44

I would thank and sorry you guys at the same time! I really meant that. Because actually, I don't know all these beforehand. I've been still constantly researching and experimenting while trying to find the best practices to apply to our application.

I think one of the factors that makes it more complex than usual is that we try to fit Clean Architecture into our new architecture. That's the reason why I'd prefer to implement a _pure MVVM_ in the first place 😄

Architecture is always opinionated and debatable. Recently I found a very interesting thread on the Apple Developer Forum that discusses [Stop using MVVM for SwiftUI](https://developer.apple.com/forums/thread/699003) which is of course, still opinionated and debatable 😄 But it has some very interesting points and insights, you guys should take a look at it 🙏

Reply:
Resolved in [bec23cc](https://github.com/nimblehq/nimble-crypto-ios/pull/75/commits/bec23cccdbbed520a37c72692a9dec37d5f77be0) 🎉